### PR TITLE
[SPARK-6390] [SQL] [MLlib] Port MatrixUDT to PySpark

### PR DESCRIPTION
**Pull request by MechCoder**
MatrixUDT was recently coded in scala. This has been ported to PySpark

### CODE REVIEW

**Comment by MechCoder** [comment]
The serialization and deserialization works but trying to create a DataFrame using a Matrix, gives me this error, 

    RuntimeError: maximum recursion depth exceeded in __instancecheck__

Code to replicate

    from pyspark.mllib.linalg import DenseMatrix, SparseMatrix, MatrixUDT
    dm1 = DenseMatrix(3, 2, [0, 1, 4, 5, 9, 10])
    sm1 = SparseMatrix(1, 1, [0, 1], [0], [2.0])
    rdd = sc.parallelize([("dense", dm1)])
    rdd.toDF()

This fails with the above mentioned error.

cc @davies @rxin Any thoughts would be appreciated,,

File: python/pyspark/mllib/linalg.py [hunk 1]
```diff
@@ -163,6 +163,59 @@ def simpleString(self):
...
+
+    @classmethod
+    def scalaUDT(cls):
+        return "org.apache.spark.mllib.linalg.MatrixUDT"
+
+    def serialize(self, obj):
+        if isinstance(obj, SparseMatrix):
+            colPtrs = [int(i) for i in obj.colPtrs]
+            rowIndices = [int(i) for i in obj.rowIndices]
+            values = [float(v) for v in obj.values]
```
Comment (davies) [inline python/pyspark/mllib/linalg.py]: Do we still need `values` for SparseMatrix, that could be HUGE.

File: python/pyspark/mllib/linalg.py [hunk 2]
```diff
@@ -163,6 +163,59 @@ def simpleString(self):
...
+            return (0, obj.numRows, obj.numCols, colPtrs,
+                    rowIndices, values, bool(obj.isTransposed))
+        elif isinstance(obj, DenseMatrix):
+            values = [float(v) for v in obj.values]
+            return (1, obj.numRows, obj.numCols, None, None, values,
+                    bool(obj.isTransposed))
+        else:
+            raise TypeError("cannot serialize %r of type %r" % (obj, type(obj)))
```
Comment (davies) [inline python/pyspark/mllib/linalg.py]: the `repr(obj)` could be very long, I think just having `type(obj)` is fine.

**Comment by davies** [comment]
@MechCoder pyUDT should be defined in MatrixUDT

**Comment by davies** [comment]
Could you also catch the exception from `_parse_datatype_json_string` in DataFrame.schema() and raise a different one (For example, just Exception()) ? Or AttributeError will be handled specially, causing infinite loop.

**Comment by MechCoder** [comment]
>  pyUDT should be defined in MatrixUDT

Thanks. I am now able to create a dataframe, but when I do `df.collect` it crashes.

**Comment by davies** [comment]
@MechCoder Could you have a unit test to reproduce the error?

**Comment by MechCoder** [comment]
Sure.

    from pyspark.mllib.linalg import DenseMatrix, SparseMatrix, MatrixUDT
    dm1 = DenseMatrix(3, 2, [0, 1, 4, 5, 9, 10])
    sm1 = SparseMatrix(1, 1, [0, 1], [0], [2.0])
    rdd = sc.parallelize([("dense", dm1)])
    df = rdd.toDF() 
    df.collect()

I know there is something silly, but I'm not able to figure out on my own, :

**Comment by davies** [comment]
@MechCoder  I had the same problem, but it turned out that there is an outdated python/lib/pyspark.zip there. After removing it, it worked fine.

**Comment by MechCoder** [comment]
@davies thanks. that worked. could you give a pass now?

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -224,7 +224,10 @@ def schema(self):
         StructType(List(StructField(age,IntegerType,true),StructField(name,StringType,true)))
         """
         if self._schema is None:
-            self._schema = _parse_datatype_json_string(self._jdf.schema().json())
+            try:
+                self._schema = _parse_datatype_json_string(self._jdf.schema().json())
+            except AttributeError:
+                raise Exception("Unable to parse datatype from schema.")
```
Comment (davies) [inline python/pyspark/sql/dataframe.py]: Could you put something about the original exception into the message?

**Comment by MechCoder** [comment]
ping @davies anything left?

**Comment by MechCoder** [comment]
jenkins retest this please

**Comment by davies** [comment]
LGTM, waiting for the tests.

**Pull request closed without merging**